MultiplayerTestRunner/src/InternalUnitTests/Test_CLI.cpp

```diff
@@ -0,0 +1,152 @@
+	EXPECT_EQ(Settings.TimeoutInSeconds, 60);
+	EXPECT_EQ(Settings.SpaceId, "space-id-123");
+}
+
+TEST(CLITest, TestIdentifierRequired)
+{
+	std::vector<char*> args = {"MultiplayerTestRunner",
+							   "--email",
+							   "test@example.com",
+							   "--password",
+							   "password123",
+							   "--space",
+							   "space-id-123",
+							   "--timeout",
+							   "60",
+							   "--endpoint",
+							   "https://example.com"};
+
+	try
+	{
+		CLIArgs::RunnerSettings Settings = CLIArgs::ProcessCLI(args.size(), args.data());
+	}
+	catch (const Utils::ExceptionWithCode& Exception)
+	{
+		EXPECT_EQ(Exception.ErrorCode, MultiplayerTestRunner::ErrorCodes::CLI_PARSE_ERROR);
+		EXPECT_EQ(std::string(Exception.what()), std::string("--test is required"));
+	}
+	catch (...)
+	{
+		FAIL() << "Unexpected exception type thrown";
+	}
+}
+
+TEST(CLITest, WhenInvalidTestIdentiferThenExceptionThrow)
+{
+	std::vector<char*> args
+		= {"MultiplayerTestRunner", "--test", "NotARealTestIdentifier", "--email", "test@example.com", "--password", "password123"};
+
+	try
+	{
+		CLIArgs::RunnerSettings Settings = CLIArgs::ProcessCLI(args.size(), args.data());
+	}
+	catch (const Utils::ExceptionWithCode& Exception)
+	{
+		EXPECT_EQ(Exception.ErrorCode, MultiplayerTestRunner::ErrorCodes::INVALID_TEST_SPECIFIER);
+		EXPECT_EQ(std::string(Exception.what()), std::string("String `NotARealTestIdentifier` does not match any TestIdentifier"));
+	}
+	catch (...)
+	{
+		FAIL() << "Unexpected exception type thrown";
+	}
+}
+
+TEST(CLITest, DefaultsSet)
+{
+	using namespace MultiplayerTestRunner::TestIdentifiers;
+	std::string TestID				 = TestIdentifierToString(TestIdentifier::CREATE_AVATAR);
+	std::vector<char*> args			 = {"MultiplayerTestRunner", "--test", TestID.data(), "--email", "test@example.com", "--password", "password123"};
+	CLIArgs::RunnerSettings Settings = CLIArgs::ProcessCLI(args.size(), args.data());
+
+	// We don't want to go to the credentials file for the defaults, so we provide a user/password
+	EXPECT_EQ(Settings.LoginEmailAndPassword, std::make_pair("test@example.com", "password123"));
+	EXPECT_EQ(Settings.TestIdentifier, TestIdentifier::CREATE_AVATAR);
+	EXPECT_EQ(Settings.Endpoint, "https://ogs-internal.magnopus-dev.cloud");
+	EXPECT_EQ(Settings.TimeoutInSeconds, 30);
+	EXPECT_FALSE(Settings.SpaceId.has_value());
+}
+
+/* This test tests different paths depending on if there's a credentials file or not */
+TEST(CLITest, WhenNoEmailThenCredentialsFileAttemptMade)
+{
+	using namespace MultiplayerTestRunner::TestIdentifiers;
+	std::string TestID		= TestIdentifierToString(TestIdentifier::CREATE_AVATAR);
+	std::vector<char*> args = {"MultiplayerTestRunner", "--test", TestID.data(), "--password", "password123"};
+
+	try
+	{
+		CLIArgs::RunnerSettings Settings = CLIArgs::ProcessCLI(args.size(), args.data());
+		// If there is a credentials file, we'll now have a login email
+		EXPECT_TRUE(Settings.LoginEmailAndPassword.first.length() > 0);
```

Comment:
[question] I'm not sure I understand this test. a password is passed as argument, but not an email, so you look for the email in the credentials file, and you use the password that was passed on the command line?
do we really want to do that? will it do that, if, e.g. the email was passed but with a typo (--emoil or similar)?

would it be better to completely decorrelate (email/password) and (creds file)? i.e. if no email and no password, then creds file. if email but no password: error, user forgot password, and vice versa.

I'm genuinely confused as to why you would use the email of one and the password of the other (it may just be me being thick)

Reply:
The idea is that if either a password or a username is not passed, it will use _both_ credentials from the credentials file, so you never use one from the CLI and one from the file, it's all or nothing.
You are right that it's a bit odd though, just from an interface perspective, not surprised it confused you. What do you think, should we make it so that you need to pass both email and password arguments or else you just get an error, rather than an automatic fallback to the credentials file?

Reply:
oh right, I didn't realise it was then using the pair from the file. it does make a lot more sense. 

my worry by falling back silently on the credentials from the file is that you may make a mistake, not realise it, and assume you are using your custom credentials when you are in fact using the default ones. it may be worth printing a warning, at least?

Reply:
https://github.com/magnopus-opensource/connected-spaces-platform/pull/558/commits/68334c43cdfb1f5cc6c96c99ba5a2e0e7ccdb9ac